In [80]:
import pandas as pd

## Initial Exploration

In [153]:
df = pd.read_csv("inventory_test.csv")
print(df.shape)
df.head()

(8454, 206)


C:\Users\pizza\AppData\Local\Temp\ipykernel_10248\2180137487.py:1: DtypeWarning: Columns (129,138,139,140,141,142,143,148,152,157,158,168,169,172,173,176,179,180,181,189,191) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("inventory_test.csv")


PatientID  PatientSex  PatientBirthDate  \
0      K100         NaN               NaN   
1      K100         NaN               NaN   
2      K100         NaN               NaN   
3      K100         NaN               NaN   
4      K100         NaN               NaN   

                                    StudyInstanceUID  \
0  1.3.12.2.1107.5.1.4.28600023236596336688906318...   
1  1.3.12.2.1107.5.1.4.28600023236596336688906318...   
2  1.3.12.2.1107.5.1.4.28600023236596336688906318...   
3  1.3.12.2.1107.5.1.4.28600023236596336688906318...   
4  1.3.12.2.1107.5.1.4.28600023236596336688906318...   

               StudyDescription  StudyDate  \
0  CT BRAIN WO CONTRAST (70450)   20220401   
1  CT BRAIN WO CONTRAST (70450)   20220401   
2  CT BRAIN WO CONTRAST (70450)   20220401   
3  CT BRAIN WO CONTRAST (70450)   20220401   
4  CT BRAIN WO CONTRAST (70450)   20220401   

                                   SeriesInstanceUID  \
0  1.3.12.2.1107.5.1.4.19743776053189314668198005...   
1  1.3.12.2.1107.5.1.4.14463949905640131236256856...   
2  1.3.12.2.1107.5.1.4.11270577611930653926640724...   
3  1.3.12.2.1107.5.1.4.17407020384920076295523451...   
4  1.3.12.2.1107.5.1.4.84337701080239475497944344...   

              SeriesDescription Manufacturer  Instances  ...  \
0           Topogram__0_6__T20s      SIEMENS          2  ...   
1       Helical_Head__5_0__J37s      SIEMENS         36  ...   
2  Helical_Head__5_0__Axial_MPR      SIEMENS         41  ...   
3          Helical_Head_Coronal      SIEMENS         49  ...   
4         Helical_Head_Sagittal      SIEMENS         49  ...   

  RETIRED_CurveDimensions  RETIRED_NumberOfPoints  RETIRED_TypeOfData  \
0                     NaN                     NaN                 NaN   
1                     NaN                     NaN                 NaN   
2                     NaN                     NaN                 NaN   
3                     NaN                     NaN                 NaN   
4                     NaN                     NaN                 NaN   

   RETIRED_CurveDescription RETIRED_DataValueRepresentation RETIRED_CurveData  \
0                       NaN                             NaN               NaN   
1                       NaN                             NaN               NaN   
2                       NaN                             NaN               NaN   
3                       NaN                             NaN               NaN   
4                       NaN                             NaN               NaN   

   SmallestPixelValueInSeries LargestPixelValueInSeries  ScanArc  \
0                         NaN                       NaN      NaN   
1                         NaN                       NaN      NaN   
2                         NaN                       NaN      NaN   
3                         NaN                       NaN      NaN   
4                         NaN                       NaN      NaN   

  RouteOfAdmissions  
0               NaN  
1               NaN  
2               NaN  
3               NaN  
4               NaN  

[5 rows x 206 columns]

In [154]:
df.dtypes

PatientID                      object
PatientSex                    float64
PatientBirthDate              float64
StudyInstanceUID               object
StudyDescription               object
                               ...   
RETIRED_CurveData             float64
SmallestPixelValueInSeries    float64
LargestPixelValueInSeries     float64
ScanArc                       float64
RouteOfAdmissions             float64
Length: 206, dtype: object

In [155]:
df.isnull().sum()

PatientID                        0
PatientSex                    8454
PatientBirthDate              8454
StudyInstanceUID                 0
StudyDescription                 0
                              ... 
RETIRED_CurveData             8454
SmallestPixelValueInSeries    8454
LargestPixelValueInSeries     8454
ScanArc                       8454
RouteOfAdmissions             8454
Length: 206, dtype: int64

In [156]:
df.isnull().any().value_counts()


True     173
False     33
dtype: int64

In [157]:
df.duplicated().sum()

0

In [158]:
df = df.dropna(axis=1, how='all')
df.shape

(8454, 178)

In [8]:
df.to_excel('inventory_test.xlsx', index=False)

## Exploring Examples.csv data

We explore the example file to achieve a deeper understanding on the information/fields of NCCT images
By doing this we create an excel file with all DICOM fields for every of the example instances by getting all ``unique`` identifiers of the ``SeriesInstanceUID`` field.

In [24]:
df2 = pd.read_csv("example.csv")
print(df2.shape)
df2.head(2)

(25, 8)


PatientID                                   StudyInstanceUID  \
0      E150  1.3.12.2.1107.5.1.4.32684195495775992931128954...   
1      E155  1.3.12.2.1107.5.1.4.23885769149733257270910067...   

                                   SeriesInstanceUID  \
0  1.3.12.2.1107.5.1.4.31168712094706013552005368...   
1  1.3.12.2.1107.5.1.4.31225837177471900865985511...   

                                    StudyDescription   SeriesDescription  \
0  CT BRAIN WO, CTA BRAIN & CEREBRAL PERFUSION (7...  Head__5_0__J37s__1   
1  CT BRAIN WO, CTA BRAIN & CEREBRAL PERFUSION (7...  Head__5_0__J37s__1   

         PixelSpacing  SliceThickness ConvolutionKernel  
0  [0.46875, 0.46875]             5.0     ['J37s', '1']  
1  [0.46875, 0.46875]             5.0     ['J37s', '1']

In [26]:
series_instances_list = df2['SeriesInstanceUID'].tolist()
print(len(series_instances_list))
print(series_instances_list)

25
['1.3.12.2.1107.5.1.4.311687120947060135520053681636376863046', '1.3.12.2.1107.5.1.4.312258371774719008659855117295968314669', '1.3.12.2.1107.5.1.4.19336894898411098690550601021425163460', '1.3.12.2.1107.5.1.4.201563371198288011919329422932203773537', '1.3.12.2.1107.5.1.4.86723521497130828333795468585801256582', '1.3.12.2.1107.5.1.4.178092708803399546524878855211014937434', '1.3.12.2.1107.5.1.4.282376521307071244316854120254008724244', '1.3.12.2.1107.5.1.4.256163344752846767866745916209868773467', '1.3.12.2.1107.5.1.4.9705880275585620621749485170460468499', '1.3.12.2.1107.5.1.4.18456855111763366437866153641678674005', '1.3.12.2.1107.5.1.4.8167708748792602775608566355216663250', '1.3.12.2.1107.5.1.4.110952798673385600635323023171560883970', '1.3.12.2.1107.5.1.4.45794954103512473475448654137847772961', '1.3.12.2.1107.5.1.4.313170610233070016233741393443607423794', '1.3.12.2.1107.5.1.4.277860558742256469027844731064080524865', '1.3.12.2.1107.5.1.4.51444219027449397256146662494276469899

In [28]:
example_df = df[df['SeriesInstanceUID'].isin(series_instances_list)]
print(example_df.shape)
example_df.to_excel('example_complete.xlsx')

(25, 178)


We have effectively created the ``example_complete.xlsx`` file to dive further into the data

# Filtering EXERCISE

The correct NCCT images must meet the following characteristics:

- Non-contrast image
- CT modality
- Axial orientation
- Slice thickness between 2.5 and 5 mm
- First (in time) NCCT acquired

Since determining the non-contrast image appears to be more challenging, as there are no specific metrics provided in the exercise or the DICOM documentation, we will first address and filter out the fields that are considered more straightforward.

Since we don't have clear metrics for defining an NCCT image, my strategy is to exclude instances that do not meet the criteria for an NCCT image.

## Filter Modality

In [159]:
df['Modality'].unique()

array(['CT', 'MR', 'OT', 'XA'], dtype=object)

In [160]:
df['ModalitiesInStudy'].unique()

array(['CT\\SR', 'CT', 'MR\\SR', 'MR', 'OT\\MR\\SR', 'OT\\CT\\SR',
       'KO\\CT', 'CT\\SR\\PR', 'OT\\KO\\CT', 'MR\\SR\\PR',
       'OT\\KO\\CT\\SR', 'OT\\CT', 'XA'], dtype=object)

In [161]:
df = df[df['Modality'] == 'CT']
df= df[df['ModalitiesInStudy'].str.contains('CT', case=False, na=False)]
df.shape

(6873, 178)

## Filter by Thickness

According to the DICOM documentation, the Slice Thickness field is in ``mm`` and it is a required value although it can be left empty if unknown.

[Reference to C.7.6.2.1 Image Plane Module Attribute Descriptions](https://dicom.nema.org/medical/dicom/current/output/chtml/part03/sect_C.7.6.2.html#table_C.7-10)

In [162]:
df = df[(df['SliceThickness'] >= 2.5) & (df['SliceThickness'] <= 5)]
df.shape

(2013, 178)

In [163]:
df['PatientID'].nunique()

325

We can see how applying conditions that are 100% clear and unambiguous the unique patient count has reduced from 326 to 325, suggesting that **not all patients may have NCCT images in this inventory file**.

In [ ]:
df.to_excel('ct_filtered.xlsx')

After applying some of the filters we have considerably reduced the size of the DataFrame from 8458 instances to just 2013. Further exploration inside the filtered Excel file gives us some insights in which columns can we have a better look at next. We use Excel since the amount of columns is fairly large and Excel offers a better UI to navigate through this data.

## Relevant Columns

In [164]:
relevant_columns_list = ['Modality', 'StudyDescription', 'SeriesDescription', 'LastUpdate', 'ConvolutionKernel', 'ImageOrientationPatient', 'SliceThickness', 'ImageType']

## Distinct values

We select some potentialy relevant columns to perform some further exploration. We will now check the distinct values for each of this fields and see if we can find a pattern to include or exclude certain rows. 

In [165]:
unique_values = {col: df[col].unique() for col in relevant_columns_list}

In [166]:
for col, unique_vals in unique_values.items():
    if col == 'LastUpdate' or col == 'ImageOrientationPatient':
        continue
    print(f"\nColumn: {col}")
    print(f"Distinct values ({len(unique_vals)}):")
    for val in unique_vals:
        print(f"   {val}")


Column: Modality
Distinct values (1):
   CT

Column: StudyDescription
Distinct values (13):
   CT BRAIN WO CONTRAST (70450)
   CT BRAIN WO, CTA BRAIN & CEREBRAL PERFUSION (70496, 0042T)
   CT ANGIO HEAD & NECK W/WO (70496, 70498)
   CTA BRAIN & CEREBRAL PERFUSION (70496, 0042T)
   EXTERNAL CT-STORE & INTERPRET
   CT ANGIO HEAD W/WO CONTRAST (70496)
   EXTERNAL CT - STORE ONLY
   EXTERNAL CT - STORE ONLY - RAD
   EXTERNAL CT BRAIN - STORE ONLY - RAD
   EXTERNAL CT BRAIN - STORE ONLY
   EXTERNAL CTA HEAD - STORE ONLY - RAD
   EXTERNAL CTA HEAD - STORE ONLY
   EXTERNAL CT BRAIN INTERPRET

Column: SeriesDescription
Distinct values (197):
   Helical_Head__5_0__J37s
   Helical_Head__5_0__Axial_MPR
   Helical_Head_Coronal
   Helical_Head_Sagittal
   Helical_Head__3_0__J70h
   Head__5_0__Hc40__2
   Bone__Head__3_0__Br64__2
   Straightened_CT_Series
   Head__5_0__Axial
   Head__5_0__Coronal
   Head__5_0__Sagittal
   Head__5_0__J37s__1
   Head__3_0__J70h__1
   Head__5_0_Axial
   Helical_Head__5

## Filter by ImageType

After closer review of the DICOM documentation 
[Reference to DICOM C.7.6.1.1.2 Image Type](https://dicom.innolitics.com/ciods/ct-image/general-image/00080008)
we decide to only include instances with the ``PRIMARY`` tag in the ImageType field. The decision to include this tag, rather than excluding the SECONDARY tag, comes from the fact that it is a required field according to the documentation.

In [167]:
df = df[df['ImageType'].str.contains('PRIMARY', case=False, na=False)]
df.shape

(1761, 178)

In [169]:
df['PatientID'].nunique()

324

According to the documentation in
[Reference to DICOM C.8.2.1.1.1 Image Type](https://dicom.innolitics.com/ciods/ct-image/ct-image/00080008)
the 3rd tag value in the ``ImageType`` field is optional and differenciates between AXIAL and LOCALIZER. Since it is an optional tag, it may or may not be included so we filter by excluding images that have the localizer tag. Pictures that belong to the localizer cathegory are not fit to train our model.

In [170]:
df = df[df['ImageType'].str.contains('localizer', case=False, na=False) == False]
df.shape

(1631, 178)

In [171]:
df['StudyDescription'].unique()

array(['CT BRAIN WO CONTRAST (70450)',
       'CT BRAIN WO, CTA BRAIN & CEREBRAL PERFUSION (70496, 0042T)',
       'CT ANGIO HEAD & NECK W/WO (70496, 70498)',
       'EXTERNAL CT-STORE & INTERPRET',
       'CT ANGIO HEAD W/WO CONTRAST (70496)',
       'CTA BRAIN & CEREBRAL PERFUSION (70496, 0042T)',
       'EXTERNAL CT - STORE ONLY', 'EXTERNAL CT - STORE ONLY - RAD',
       'EXTERNAL CT BRAIN - STORE ONLY - RAD',
       'EXTERNAL CT BRAIN - STORE ONLY',
       'EXTERNAL CTA HEAD - STORE ONLY - RAD',
       'EXTERNAL CT BRAIN INTERPRET', 'EXTERNAL CTA HEAD - STORE ONLY'],
      dtype=object)

## Filter by Axial Pictures

We need to define a way to differenciate further between axial images and non-axial images. The documentation suggests that the orientation should be [1, 0, 0, 0, 1, 0] although judging from the example.csv file where instances of some NCCT images are displayed, there seems to be some small tolerance threshold to allow for small angle deviations. 

In [172]:
df[['X_x', 'X_y', 'X_z', 'Y_x', 'Y_y', 'Y_z']] = df['ImageOrientationPatient'].str.split('\\', expand=True).astype(float)

In [173]:
def is_axial(row, tolerance=0.06):

    if 'axial' in row['ImageType'].lower():
        return True
    
    return (
        abs(row['X_x'] - 1) <= tolerance and
        abs(row['X_y']) <= tolerance and
        abs(row['X_z']) <= tolerance and
        abs(row['Y_x']) <= tolerance and
        abs(row['Y_y'] - 1) <= tolerance and
        abs(row['Y_z']) <= tolerance
    )

In [174]:
df['IsAxial'] = df.apply(lambda row: is_axial(row, tolerance=0.06), axis=1)
df.shape

(1631, 185)

In [175]:
print(df[['ImageOrientationPatient', 'IsAxial']].head())

                             ImageOrientationPatient  IsAxial
1                                        1\0\0\0\1\0     True
2  0.99722881968274\-0.074395437993\0\0.058872430...     True
3  0.99567980109701\-0.092853291204\0\-0.04248384...     True
4  0.07417242893302\0.74222162601923\-0.666037167...     True
5                                        1\0\0\0\1\0     True


In [176]:
df['IsAxial'].value_counts()


True     1608
False      23
Name: IsAxial, dtype: int64

We only include those instances where the IsAxial field is ``True``

In [177]:
df = df[df['IsAxial'] == True]
df.shape

(1608, 185)

In [178]:
df['StudyDescription'].unique()

array(['CT BRAIN WO CONTRAST (70450)',
       'CT BRAIN WO, CTA BRAIN & CEREBRAL PERFUSION (70496, 0042T)',
       'CT ANGIO HEAD & NECK W/WO (70496, 70498)',
       'EXTERNAL CT-STORE & INTERPRET',
       'CT ANGIO HEAD W/WO CONTRAST (70496)',
       'CTA BRAIN & CEREBRAL PERFUSION (70496, 0042T)',
       'EXTERNAL CT - STORE ONLY', 'EXTERNAL CT - STORE ONLY - RAD',
       'EXTERNAL CT BRAIN - STORE ONLY - RAD',
       'EXTERNAL CT BRAIN - STORE ONLY',
       'EXTERNAL CTA HEAD - STORE ONLY - RAD',
       'EXTERNAL CT BRAIN INTERPRET', 'EXTERNAL CTA HEAD - STORE ONLY'],
      dtype=object)

In [76]:
df_filtered2 = df 
df_filtered2.to_excel('filtered_more.xlsx')

We save again our filtered DataFrame to continue further exploration in Excel

## Distinct values exploring

We can see a certain pattern in the SeriesDescription field and ImageType

In [179]:
df_filtered2 = df
df_filtered2.shape

(1608, 185)

In [180]:
relevant_columns_list2 = ['Modality', 'StudyDescription', 'SeriesDescription', 'ImageType', 'BodyPartExamined']

In [181]:
unique_values2 = {col : df_filtered2[col].unique() for col in relevant_columns_list2}



In [182]:
for col, unique_vals2 in unique_values2.items():
    print(f"\nColumn: {col}")
    print(f"Distinct values ({len(unique_vals2)}):")
    for val in unique_vals2:
        print(f"   {val}")



Column: Modality
Distinct values (1):
   CT

Column: StudyDescription
Distinct values (13):
   CT BRAIN WO CONTRAST (70450)
   CT BRAIN WO, CTA BRAIN & CEREBRAL PERFUSION (70496, 0042T)
   CT ANGIO HEAD & NECK W/WO (70496, 70498)
   EXTERNAL CT-STORE & INTERPRET
   CT ANGIO HEAD W/WO CONTRAST (70496)
   CTA BRAIN & CEREBRAL PERFUSION (70496, 0042T)
   EXTERNAL CT - STORE ONLY
   EXTERNAL CT - STORE ONLY - RAD
   EXTERNAL CT BRAIN - STORE ONLY - RAD
   EXTERNAL CT BRAIN - STORE ONLY
   EXTERNAL CTA HEAD - STORE ONLY - RAD
   EXTERNAL CT BRAIN INTERPRET
   EXTERNAL CTA HEAD - STORE ONLY

Column: SeriesDescription
Distinct values (117):
   Helical_Head__5_0__J37s
   Helical_Head__5_0__Axial_MPR
   Helical_Head_Coronal
   Helical_Head_Sagittal
   Helical_Head__3_0__J70h
   Head__5_0__Hc40__2
   Bone__Head__3_0__Br64__2
   Head__5_0__Axial
   Head__5_0__Coronal
   Head__5_0__Sagittal
   Head__5_0__J37s__1
   Head__3_0__J70h__1
   Head__5_0_Axial
   Helical_Head__5_0__H41s
   Bone_Head__3_0

## Filtering by SeriesDescription

The SeriesDescription field offers some clues on how to further filter out the data. We include those instances where the word ``head`` is contained within the ``SeriesDescription`` field.

In [183]:
df = df[df['SeriesDescription'].str.contains('head', case=False, na=False)]
df.shape

(1543, 185)

In [184]:
df['StudyDescription'].unique()

array(['CT BRAIN WO CONTRAST (70450)',
       'CT BRAIN WO, CTA BRAIN & CEREBRAL PERFUSION (70496, 0042T)',
       'CT ANGIO HEAD & NECK W/WO (70496, 70498)',
       'EXTERNAL CT-STORE & INTERPRET',
       'CT ANGIO HEAD W/WO CONTRAST (70496)',
       'CTA BRAIN & CEREBRAL PERFUSION (70496, 0042T)',
       'EXTERNAL CT - STORE ONLY', 'EXTERNAL CT - STORE ONLY - RAD',
       'EXTERNAL CT BRAIN - STORE ONLY - RAD',
       'EXTERNAL CT BRAIN - STORE ONLY', 'EXTERNAL CT BRAIN INTERPRET'],
      dtype=object)

## Filtering by StudyDescription

We notice the unique values on the StudyDescription field have shrinked. Upon further research, we decide to exclude instances that contain ``ANGIO`` on the StudyDescription field as these refer to images with contrast, which is the opposite of what we want.

In [185]:
df = df[~df['StudyDescription'].str.contains('ANGIO', case=False, na=False)]
df.shape

(1472, 185)

In [186]:
df['StudyDescription'].unique()

array(['CT BRAIN WO CONTRAST (70450)',
       'CT BRAIN WO, CTA BRAIN & CEREBRAL PERFUSION (70496, 0042T)',
       'EXTERNAL CT-STORE & INTERPRET',
       'CTA BRAIN & CEREBRAL PERFUSION (70496, 0042T)',
       'EXTERNAL CT - STORE ONLY', 'EXTERNAL CT - STORE ONLY - RAD',
       'EXTERNAL CT BRAIN - STORE ONLY - RAD',
       'EXTERNAL CT BRAIN - STORE ONLY', 'EXTERNAL CT BRAIN INTERPRET'],
      dtype=object)

Some closer look on the ``StudyDescription`` field from the original ``inventory_test.csv`` file, gives us some more clues on how the values inside this field are organized. 
Sometimes the value contained belongs to **more than one type of study**, these type of studies are separated by commas. Another important thing to take into account is that sometimes, the values also include a refered code i.e. 
> (70496, 0042T) 

We need to also take that into account. We want to **exclude** those values where there is only one type of study and the word CTA is included. 
Since we cannot create a rule to include ALL NCCT instances this would be our best best to not roll out any possible NCCT images. 

In [187]:
# Exclude rows where 'CTA' is present and:
# 1. No commas are present, or
# 2. The comma is preceded by a number and followed by a space
df = df[~(df['StudyDescription'].str.startswith('CTA') & 
                   df['StudyDescription'].str.count(',').eq(1) & 
                   df['StudyDescription'].str.contains(r'\d,\s'))]

df.shape

(1471, 185)

In [188]:
df['StudyDescription'].unique()

array(['CT BRAIN WO CONTRAST (70450)',
       'CT BRAIN WO, CTA BRAIN & CEREBRAL PERFUSION (70496, 0042T)',
       'EXTERNAL CT-STORE & INTERPRET', 'EXTERNAL CT - STORE ONLY',
       'EXTERNAL CT - STORE ONLY - RAD',
       'EXTERNAL CT BRAIN - STORE ONLY - RAD',
       'EXTERNAL CT BRAIN - STORE ONLY', 'EXTERNAL CT BRAIN INTERPRET'],
      dtype=object)

In [189]:
unique_values3 = {col : df[col].unique() for col in relevant_columns_list2}

for col, unique_vals3 in unique_values3.items():
    print(f"\nColumn: {col}")
    print(f"Distinct values ({len(unique_vals3)}):")
    for val in unique_vals3:
        print(f"   {val}")


Column: Modality
Distinct values (1):
   CT

Column: StudyDescription
Distinct values (8):
   CT BRAIN WO CONTRAST (70450)
   CT BRAIN WO, CTA BRAIN & CEREBRAL PERFUSION (70496, 0042T)
   EXTERNAL CT-STORE & INTERPRET
   EXTERNAL CT - STORE ONLY
   EXTERNAL CT - STORE ONLY - RAD
   EXTERNAL CT BRAIN - STORE ONLY - RAD
   EXTERNAL CT BRAIN - STORE ONLY
   EXTERNAL CT BRAIN INTERPRET

Column: SeriesDescription
Distinct values (74):
   Helical_Head__5_0__J37s
   Helical_Head__5_0__Axial_MPR
   Helical_Head_Coronal
   Helical_Head_Sagittal
   Helical_Head__3_0__J70h
   Head__5_0__Hc40__2
   Bone__Head__3_0__Br64__2
   Head__5_0__Axial
   Head__5_0__Coronal
   Head__5_0__Sagittal
   Head__5_0__J37s__1
   Head__3_0__J70h__1
   Head__5_0_Axial
   Helical_Head__5_0__H41s
   Bone_Head__3_0__H70h
   Infant_Head__4_0__Hr38__2
   Infant_Head__4_0__Axial
   Head__5_0__J37f__1
   Head__3_0__Q40f__1
   Head__5_0__MPR__ax
   Head__5_0__MPR__cor
   Head__5_0__MPR__sag
   Head___5_0__Hc40__2
   Head___

In [286]:
df.to_excel('final_filtered.xlsx', index=False)

## Creating a DateTime column

In [190]:
print(df[['ContentDate', 'ContentTime']].dtypes)

ContentDate    float64
ContentTime    float64
dtype: object


In [191]:
null_values = df[['ContentDate', 'ContentTime']].isna().sum()

print(null_values)

ContentDate    0
ContentTime    0
dtype: int64


We have made sure that no null values are present on either of our Date or Time fields, note that we need to convert it from float to string in order to work later with datetime.
The ContentTime field contains some faulty data since the leading zero in the time is missing so for example:

9:54:00 a.m. is written as 95400 but we need it to be 095400, so we will add a leading 0 for those instances

In [192]:
df['ContentDate'] = df['ContentDate'].astype(float).astype(int).astype(str)
df['ContentTime'] = df['ContentTime'].astype(float).astype(int).astype(str)

df['ContentTime'] = df['ContentTime'].str.zfill(6)

In [193]:
df['ContentDateTime'] = pd.to_datetime(df['ContentDate'] + df['ContentTime'], format='%Y%m%d%H%M%S')

## Grouping by PatientID and Ordering by ContentDateTime

We sort values by PatientID AND our newly created column ContentDateTime. We will pick only the first in time (the oldest one).

In [194]:
df_sorted = df.sort_values(by=['PatientID', 'ContentDateTime'])

df_first_instance = df_sorted.groupby('PatientID').first().reset_index()

In [195]:
df_first_instance.shape

(320, 186)

We are left wit 320 patients, this could mean that our filtering strategy has been too aggressive or that as we supposed at the beginning, not all PatientsID contain NCCT images. 

In [196]:
df['PatientID'].nunique()

320

## That's all folks
We save the final result with the unique patients and the first image in time

In [152]:
df_first_instance.to_excel('final_result.xlsx', index=False)

## Notes:

- Since it is specified that the values within the 25 instances are 100% NCCT images, I had to assume that there was some tolerance within the orientation inside the ``ImageOrientationPatient`` as there was a row where the ``ImageType`` field did not contain the  ``AXIAL`` tag and the orientation vector was not perfectly aligned with [1, 0, 0, 0, 1, 0], which made me think that a tolerance value was being used.

- In the inventory, in particular in the ``BodyPartExamined`` field I have found that it contains values that are not within the official code denomination [Table L-1. Corresponding Codes and Terms for Human Use](https://dicom.nema.org/medical/dicom/current/output/chtml/part16/chapter_L.html#chapter_L) i.e. CT_HEAD, CT_ HEAD